### ISSUE TITLE
Discuss a selection of "default" Jupyter extensions in Anaconda

### ISSUE THREAD

**Issue by ijstokes** [jupyter/jupyter#124]
Anaconda (http://continuum.io/download) is a major avenue for people to get started with Jupyter.  We, at Continuum, believe Jupyter lowers the bar significantly for a wide range of people who otherwise wouldn't have access to the tools and interfaces Jupyter presents.  We observe two important things about this "new user community" (in contrast to "power users"):
- They get updates of Jupyter by downloading our quarterly major release of Anaconda
- They never want to do anything at the command line, and certainly no operations that they perceive may break their Anaconda (and by implication: Jupyter) software. This means:
  - They won't do `conda install` or `pip install` or `git anything`
  - They won't do `conda create` or `virtualenv venv`
  - They will be very hesitant doing any _bang_ operations inside of Jupyter

Our response to this situation is that we want to provide even more "out of the box" goodness by default (extending the idea of the rich set of Conda packages we already provide -- over 200), and we'd like to include ~half-dozen notebook extensions pre-installed, pre-configured, and pre-enabled with the next release of Anaconda in mid/late Spring 2016.  We expect this will ship with Jupyter 4.2 (or possibly 4.3 if it is complete by the change-freeze date for the Anaconda Release -- typically 2 weeks prior to target release).

We want to make sure everyone understands that "virgin" Jupyter would always be available to Anaconda users by at least these two methods:
- miniconda + `conda install jupyter`
- `conda create -n vanilla_jupyter jupyter`

Which is to say, the Jupyter package we create will continue to be "just vanilla Jupyter + its core dependencies"

We would like to get input on:
- Is the Jupyter community generally supportive of this idea?
- Are there any considerations we need to think about over the next 4-8 weeks leading up to this?
- Which extensions should we include.

On the last point, we'd like to start with a fairly short and conservative set (perhaps a half dozen extensions), and focus on selecting extensions from these two sets:
- Extensions already tracked here: https://github.com/ipython-contrib/IPython-notebook-extensions#documentation
- Extensions that provide a more cohesive user experience for the Anaconda user, providing Jupyter-internal GUI elements for conda-related operations, access to Anaconda Cloud, and creation of rich presentations.

Continuum's selection criteria right now (but these are open for discussion!) are to bias towards the small set of extensions that are:
- generally applicable (most Jupyter users will find them useful)
- robust (well tested, widely used)
- cross-platform (Windows, Mac, Linux)
- minimal or no dependencies beyond what Anaconda already provides or can be quickly, easily, and reliably fetched or shrink-wrapped (NPM packages).
- utilize the 4.2 extension API for uniform extension deployment, configuration, and enable/disable.

**Comment by Carreau** [comment]
From JupyterDays feedbacks, toggling and persisting all lines number on codecell + a toolbar button would be appreciated:

https://github.com/ipython-contrib/IPython-notebook-extensions/blob/master/nbextensions/usability/toggle_all_line_numbers/main.js is a start, but would need some refactoring. 

With a bit of work, it **should** be possible to have an extension that allow to edit (some) keybindings:

which I'm trying to push into master : 

https://github.com/jupyter/notebook/pull/1234

But this does not affect CodeMirror.

@parleur is working on making a the keymap of CodeMirror switchable:

https://github.com/jupyter/notebook/pull/1109

The vim mode is tricky, but he had default/emacs/sublime working. Maybe he would be able to develop that as a pure extension for 4.x

**Comment by merqurio** [comment]
I would like to propose [CodeMirror Syntax Highlight](https://github.com/merqurio/jupyter_themes) theming. Per se, It just modifies the code cells CodeMirror's theme and stores the theme in the notebooks settings, making the selected theme persisting.

I also want to add the possibility of modifying some other CodeMirror configurations like showing line number or font. 

I'm open to suggestions or modifications to make it a default extension. This was originally created to make clear to newcomers which cells where code and which where plain text and it had a great feedback.

**Comment by janschulz** [comment]
A [table of content extension](https://github.com/ipython-contrib/IPython-notebook-extensions/tree/master/nbextensions/usability/toc2) is very helpful if you have a big notebook.

**Comment by parente** [comment]
You might consider some of the incubator projects on the jupyter/roadmap. All of these are compatible with Notebook 4.0.x through 4.2.x.

- https://github.com/jupyter-incubator/dashboards extension lets users arrange notebook widgets/outputs in grid- or report- like dashboards, and persists these layouts into notebook documents for easy viewing by other users of the Notebook extension

- https://github.com/jupyter-incubator/declarativewidgets enables the use of Polymer web components in Python, R, and Scala with elements for binding to kernel-side events and data structure

- https://github.com/jupyter-incubator/contentmanagement adds full-text search over notebooks, the ability to import Python notebooks as modules with rich help, the ability to associate files with notebooks, download them as a bundle, and other content management additions

There's also https://github.com/jupyter-incubator/dashboards_bundlers which enable the deployment of dashboard notebooks as standalone web apps, but these are more reference implementations at the moment, and depend on deployment targets like https://github.com/jupyter-incubator/dashboards_server

Full disclosure: I'm one of the devs. /cc @lbustelo

**Comment by ijstokes** [comment]
We're working towards Anaconda 4.1 now and I'm hoping to get a short list of extensions to include.  Continuum-created extensions that will probably make it in:

* Conda Kernels: Select from different available kernels in the current Conda environment
    * https://github.com/Anaconda-Platform/nb_conda_kernels
* Conda Tab: manage Conda packages and environments from the File Browser view and install packages into the current environment from within the Notebook Editor view.
    * https://github.com/Anaconda-Platform/nb_conda
* NB Present: Compose presentations from Notebook cells (input and output):
    * https://github.com/Anaconda-Platform/nbpresent
* Anaconda Cloud: links Jupyter to Anaconda Cloud for publishing/sharing
    * https://github.com/Anaconda-Platform/nb_anacondacloud/ 

The community extensions I'd like to include (to be clear, at this point we don't yet have internal consensus on how we'd like to move forward, so these are still my personal views):

* IPyWidgets ( @SylvainCorlay): Interactive widgets for Notebooks
    * https://github.com/ipython/ipywidgets
* Dashboards ( @parente ): dashboard construction from a Notebook
    * https://github.com/jupyter-incubator/dashboards
* Content Management ( @parente ): search, import, bundling
    * https://github.com/jupyter-incubator/contentmanagement

Others in the running:

* ipyleaflet ( @ellisonbg ) https://github.com/ellisonbg/ipyleaflet
* pythreejs ( @jovyan) https://github.com/jovyan/pythreejs
* pivot table ( @nicolaskruchten) https://github.com/nicolaskruchten/pivottable
* declarative widgets ( @parente )

Of course we'll consider others both now and in the future, and I think the "selection criteria" (5 bullet points at bottom of issue description) will largely remain the same.

Questions this raises:

* Should these all be "enabled" with Anaconda installation?
* Will they create too much clutter in the menu bar?
* Will any of them conflict at a code level?
* Will this introduce confusion, especially if there is functionality overlap (e.g. two widget extensions, two layout refinement extensions)

I would love to say that we could simply include all the extensions in https://github.com/ipython-contrib/IPython-notebook-extensions however we need to have a model that is compatible with conda-based sandboxing of extension configurations and provides a set of stable cross-platform extensions.  I think that trying to include all of those (and the extension configuration system) would be biting off more than we could chew at this point.

**Comment by SylvainCorlay** [comment]
As of now, we don't have a means to have the enabling of extension enable their dependencies. Since `widgetsnbextension` (the extension for ipywidgets) is a dependency to many, it should probably be enabled by default.

**Comment by SylvainCorlay** [comment]
Other question: what about `bqplot`?

**Comment by lbustelo** [comment]
@ijstokes Please put my name next to `declarative widgets`. Also, would like to see `declarative widgets` more towards the yes pile than the maybe pile 😄 .

**Comment by bollwyvl** [comment]
> As of now, we don't have a means to have the enabling of extension enable their dependencies

So, our current best-effort solution to this (in the face of still not having `conf.d`-style configuration) is to have a metapackage, `anaconda-nb-extensions-config-meta`, that populates `$PREFIX/etc/jupyter` in its `build`.

This is gross, but is currently the only way to get the effect without making the user do a ton of extra work.

When the distribution is actually built, post-link scripts just wouldn't run.

> Should these all be "enabled" with Anaconda installation?

This goes back to my earlier suggestion of actually adding something in the UI that can access the `~/.jupyter` env... the ipython-extensions approach, but expanded to cover the rest of the API. 

> Will they create too much clutter in the menu bar?

As to UI overload: won't know until we try it! 

> Will any of them conflict at a code level?

No way to try this either until we have a fully-populated environment with all bells and whistles.

Having a robust integration test matrix across all these things, in real browsers, is the only way to assure this at a high level.

**Comment by bollwyvl** [comment]
> IPyWidgets (@SylvainCorlay): Interactive widgets for Notebooks

[ipywidgets](https://github.com/conda-forge/ipywidgets-feedstock), and its dependencies, are already built on conda-forge.

To that end, IMHO, I think conda-forge is the right "incubator" for getting these types of projects into the distribution: I was able to enable cookiecutter and its dependencies as a [single PR](https://github.com/conda-forge/staged-recipes/pull/631) (somewhat discouraged, as it creates non-euclidean build queue issues), and now the recipes will be easy to maintain in the future as new versions come out. Indeed, before versions come out, and "daily" `-c conda-forge` users would already be testing out versions, so we could see how things are developing.

So basically, the gating for a new package would be:
- get your stuff building on conda-forge!
- submit a PR to the metapackage!
- discuss!

Then, when the window rolls around for an anaconda distribution release, freeze the addition of new packages.

**Comment by bollwyvl** [comment]
> @ijstokes Please put my name next to declarative widgets. Also, would like to see declarative widgets more towards the yes pile than the maybe pile 😄 .

@lbustelo: DW has a hard runtime requirement on a global install of [`bower`](https://github.com/jupyter-incubator/declarativewidgets/blob/bfa54facded6bef122abcc98fc98faafe56436f4/nb-extension/python/urth/widgets/ext/urth_import.py#L19) (t/f `nodejs`), right?

There is an existing [bower package on the `javascript` channel](https://anaconda.org/javascript/bower), but we would need to get our ducks in order on how that actually gets built and kept up-to-date. Presumably https://github.com/conda-forge/bower-feedstock would have to become the package we are looking for.

That being said, while `nodejs` is currently "in" anaconda (available from `defaults`), it is not in the distribution... that might be a showstopper until something more mainline (like the notebook itself) uses node at runtime... but I could be totally off-base.

**Comment by Carreau** [comment]
That seem like an acceptable list to me.

**Comment by SylvainCorlay** [comment]
as long as bqplot is in it ;)

**Comment by nicolaskruchten** [comment]
I'd be super-happy to help get my pivot table functionality in! What's the best way for me to do so?

**Comment by lbustelo** [comment]
@bollwyvl The `bower` requirement comes into play if you want to allow the user to install webcomponents from the web. We could always pre-package a bunch of the polymer catalog in.

**Comment by bollwyvl** [comment]
> @SylvainCorlay as long as bqplot is in it ;)

It would make it much easier to write matplotlib vs seaborn vs bokeh vs bqplot comparisons :) I see no reason for it not to be included.

> I'd be super-happy to help get my pivot table functionality in! What's the best way for me to do so?

@nicolaskruchten great functionality! I think I even did an implementation with pivottablejs a while back, but with widgets... dataframes are obviously a better place to start :)

I think a rule of thumb in anaconda distribution is that a package shouldn't have run-time dependencies on anything not in the distribution. Of course there are exceptions, like like multi-GB natural language corpora, but for some web assets we should be able to pack them all up nicely.

Hence, your iframe integration approach would probably need to have all of those CDNJS links replaced with a node build stack, and then have those assets installed via normal `nbextension` mechanisms, etc.

**Comment by bollwyvl** [comment]
> The bower requirement comes into play if you want to allow the user to install webcomponents from the web. We could always pre-package a bunch of the polymer catalog in.

@lbustelo Ah, thanks for the clarification. I think having a suitable library of stuff prepacked, at your discretion, and then allowed a user to opt in to the whole node stack would be a fine way to proceed. Of course, it would be ideal to direct said user to a convenient way to do the install of bower, so we may want to make a conda-forge feedstock for `bower` anyway.

**Comment by nicolaskruchten** [comment]
@bollwyvl ok, I finally have some time to devote to PivotTable.js/Jupyter integration work... Are there any resources anywhere I could read in order to find out how to "have all of those CDNJS links replaced with a node build stack, and then have those assets installed via normal nbextension mechanisms, etc" ? I'm having troubling finding a starting point... Perhaps a similar extension that I could base my work on?

I'm happy to correspond by email as well!

**Comment by janschulz** [comment]
BTW: @jcb91split off the nb extension config page: https://github.com/jcb91/jupyter_nbextensions_configurator This thing makes the enabling of additional frontend extensions a breeze, so maybe add that to the mix?

**Comment by bollwyvl** [comment]
@nicolaskruchten here's the main doc for the API part, which outlines what has to happen:
http://jupyter-notebook.readthedocs.io/en/latest/extending/frontend_extensions.html

The build itself can be however you want... we recently added some officially-supported automation with [widget-cookiecutter](https://github.com/jupyter/widget-cookiecutter) that uses `webpack`. It is not exactly what you are looking for for your base case (just drop  some HTML/JS), but might actually be a good way to go, as the widget folks are resolving some of the long-standing embedding issues related to widgets... and would let people easily get things like selections back out of the widgets.

Anyhow, the doc there is pretty reasonable, and you could reuse a lot of what gets generated without actually making a widget.

To keep the current "here's some HTML", something like `bokeh` uses some special commands, like `output_notebook`, that would do the asset embedding, basically making the namespace available.

Indeed, happy to work offline or 1-on-1 to get something put together!

**Comment by bollwyvl** [comment]
@jcb91 https://github.com/jcb91/jupyter_nbextensions_configurator looks great! I'll have to look into it more!

**Comment by jcb91** [comment]
@bollwyvl sure 😄 It's now probably better to use the version at `jupyter-contrib`:  https://github.com/Jupyter-contrib/jupyter_nbextensions_configurator

**Comment by jcb91** [comment]
(they're the same, just that'll be the 'official' one in time. Currently pypi still points to my copy)

**Comment by ivanov** [comment]
Hey there, I'm going through old issues and it seems to me that it makes sense to close this one, as there hasn't been any new activity here for 8 years.

Thanks everyone and happy hacking! :bowtie:
